### PR TITLE
UPDATE_APPEARANCE PROFILING TESTMERGE

### PR DESCRIPTION
**Pull request by LemonInTheDark**
## About The Pull Request

[Saves 0.2 seconds of init time. 50% of emissive blockers](https://github.com/tgstation/tgstation/commit/8318b648f6d32844aacbfb4c309152cd45801f5c)

Emissive blockers are a decent expense during init, even these, which are the ones that update outside of initialize.
I've inlined them, removed some redundant vars and checks, reduced the arg count, and shifted some things around. This ends up saving 200ms, or 50% of its total cost.

I also shifted mutable_appearance about a bit. it's not a massive saving, but it is technically faster

[Prevents a few redundant appearance_updates, saves 0.8 seconds of init](https://github.com/tgstation/tgstation/commit/5475cd778b66b22b1e2c8d86b2c6d59fb84f219a)

Prequisit info: update_appearance is decently expensive
It's good then to only do it if we have a reason to, right?

Me and moth were shooting the shit about just general init time, and we came up with the idea of tracking which update_appearances actually "worked" and which didn't.

That bit comes later, let's enjoy the fruits of that work first

First, holograms were calling update_appearance on process, for almost no reason. 
I patched the one event they don't already "react" to, and then locked it behind a change decection if. 
good for live, doesn't impact init.

Next, decals. If you add a decal to something before it inits, it'll react to the after successful init signal. 
The trouble is the same atom could have its appearance updated from this MORE then once, since decals can be stacked on tiles, and signal unregisters don't work once the signal is sent.
So we add a flag to track if we've had this happen to us or not, so it only happens once. 
saves 80 ms

Power! lots of things call power_change on init, often more then once. 
We'll update appearance for each of those calls, even if only one is an actual change.
That's silly, better to track what sort of power we're using for our appearance and go off that changing

This was taking about 300ms. Really stupid

Icon smoothing. After emissive blockers were added, any change to something's icon smoothing would lead to an update_appearance call.
Nasty shit, specially cause of walls, which don't even use emissive blockers.
Ok then, so we'll always update appearance for movables, and will allow turfs that are interested to hook it manually. 
Not many of those anyhow
This is slightly a dev ux thing, but it saves 600ms so I think it's worth it. Rare case anyway

Telecomms:
telecomm machines were updating appearance on process. This is to cover for them turning on/off on process.
Better then to just check if on actually changed. 
This cost adds up midgame, doesn't impact init tho

Materials:
There's this update_appearance call in material on_apply. it doesn't do anything.
The logs will lie to you and say it does, but it's just like reapplying emissives. It doesn't need to exist
Saves like 50ms

[Uses defines to wrap update_appearance for tracking](https://github.com/tgstation/tgstation/commit/4fa82e1c9d93577aadb3c743f17196331f62e67c)

I need file and line number for my tracking, so I need to override update_appearance calls, and also preferably not require every override of update_appearance to handle dummy file + line args.

So instead, I created a wrapper proc that checks to see if appearanaces match (they're unique remember, the two of the same visual appearance will be equivalent)
The trouble is I can't intercept JUST proc calls, or JUST function definitions with defines. it needs to be both.

So I renamed the /update_appearance proc to /_update_appearance

this way I can capture old uses, and don't need to worry about merge/dev brain skew

It does mean that all update_appearance proc definitions now look weird tho. 
My profiling is leaking into dev ux. I wish I had better templating.

## Why It's Good For The Game

Saves init time, saves machine cpu usage, allows for appearance profiling in future

### CODE REVIEW

**Comment by Mothblocks** [comment]
We probably should not be merging this if you have to `_update_appearance`, can we just test merge this and get intel instead?

File: code/__HELPERS/stat_tracking.dm [hunk 1]
```diff
@@ -25,3 +27,15 @@
...
+
+/proc/stat_tracking_export_to_csv_later(filename, costs, counts)
+	if (IsAdminAdvancedProcCall())
+		return
+
+	var/list/output = list()
+
+	output += "key, cost, count"
+	for (var/key in costs)
+		output += "[key], [costs[key]], [counts[key]]"
```
Comment (Mothblocks) [inline code/__HELPERS/stat_tracking.dm:39]: You should probably validate (just throwing an exception is fine) that `key` doesn't have any commas

**Comment by Mothblocks** [comment]
Idea: Include the appearance tracking code in master today, but leave it completely unused.

Include a comment explaining the grep you need to find and replace with in order to make it work, and what defines to flip, etc.

That way we can do this again in a year without needing to review the code as finely.

**Comment by LemonInTheDark** [comment]
> Idea: Include the appearance tracking code in master today, but leave it completely unused.
> 
> Include a comment explaining the grep you need to find and replace with in order to make it work, and what defines to flip, etc.
> 
> That way we can do this again in a year without needing to review the code as finely.

I'm down, sounds like it'd work ok without interrupting dev ux

**Comment by Mothblocks** [comment]
Let's make a PR just for that, and maybe combine it with the emissive blocker micro-opts, so we have something to merge into master. Then you can keep this PR up with a DNM.

**Comment by LemonInTheDark** [comment]
> Screenshot tests failed!
> 
> [Download zip file of new screenshots.](https://file.house/WXN8.zip)
> ## Diffs
> See snapshot diffs
> Name 	Expected image 	Produced image 	Diff
> screenshot_antag_icons_abductor 	 	 	
> screenshot_antag_icons_blobinfection 	 	 	
> screenshot_antag_icons_bloodbrother 	 	 	
> screenshot_antag_icons_changeling 	 	 	
> screenshot_antag_icons_clownoperative 	 	 	
> screenshot_antag_icons_cultist 	 	 	
> screenshot_antag_icons_fugitive 	 	 	
> screenshot_antag_icons_headrevolutionary 	 	 	
> screenshot_antag_icons_heretic 	 	 	
> screenshot_antag_icons_hereticsmuggler 	 	 	
> screenshot_antag_icons_loneoperative 	 	 	
> screenshot_antag_icons_nightmare 	 	 	
> screenshot_antag_icons_obsessed 	 	 	
> screenshot_antag_icons_operative 	 	 	
> screenshot_antag_icons_operativemidround 	 	 	
> screenshot_antag_icons_provocateur 	 	 	
> screenshot_antag_icons_spaceninja 	 	 	
> screenshot_antag_icons_syndicateinfiltrator 	 	 	
> screenshot_antag_icons_syndicatesleeperagent 	 	 	
> screenshot_antag_icons_traitor 	 	 	
> screenshot_antag_icons_wizard 	 	 	
> screenshot_antag_icons_wizardmidround 	 	 	
> ## Help
> What is this?
> I am changing sprites, it's supposed to look different.
> This is a false positive.

image
works on my machine?

**Comment by LemonInTheDark** [comment]
TMing this, will open a mergable pr in a sec

File: code/__HELPERS/stat_tracking.dm (hunk 1 above)
Comment (LemonInTheDark) [reply code/__HELPERS/stat_tracking.dm:39]: gonna just strip all commas, see new pr

**Comment by LemonInTheDark** [comment]
bro wtf

**Comment by LemonInTheDark** [comment]
looking into it

**Comment by LemonInTheDark** [comment]
miscalled update_appearance in the airlock mapping helper. I've fixed it.

**Pull request closed without merging**